### DISCUSSION TITLE
Switch from Windows to Ubuntu TPM PCR0 Reconstruction failure

### DISCUSSION THREAD

**Discussion by AndreasSchaffhauser** [fwupd/fwupd#3062] [Q&A]
Dear all,

I need some help after a switch from Windows to Ubuntu 20.04.1 LTS: I've bought a Lenovo T15P with a pre-installed Windows 10. After the deletion of Windows 10 and the installation of Ubuntu 20.04.1, I want to update the firmware using fwupd.

fwupdmgr get-devices gives me at one point the following outpout:
 ─System Firmware:
│     Device ID:           5f177a3bd519ece03c345228af4255a1f5916ae2
│     Current version:     0.1.19
│     Vendor:              LENOVO (DMI:LENOVO)
│     Update Error:        TPM PCR0 differs from reconstruction, please see https://github.com/fwupd/fwupd/wiki/TPM-PCR0-differs-from-reconstruction
│     GUID:                e1678770-bdc9-47f9-8d90-340e28b6f196
│     Device Flags:        • Internal device
│                          • Updatable
│                          • Requires AC power
│                          • Supported on remote server
│                          • Needs a reboot after installation
│                          • Cryptographic hash verification is available
│                          • Device is usable for the duration of the update

It seems it's the UEFI which causes the TPM PCR0 failure. The current version of the UEFI of the site from the Lenovo vendor site is 1.19 (see https://support.lenovo.com/us/en/downloads/ds546302-bios-update-utility-bootable-cd-for-windows-10-64-bit-linux-thinkpad-p15-gen-1-p17-gen-1-t15g-gen-1-t15p-gen-1-p15v-gen-1)

Trying to use the normal workflow fwupdmgr get-devices, fwupdmgr refresh, fwupdmgr get-updates, fwupdmgr update the problem still persists. Could be here the problem, that before my switch to Ubuntu the current Version of UEFI 1.19 was installed in a "windows manner", i.e. with Lenovo Vantage or is it a problem with the fwupd version?

The current version of my fwupd is the following:
client version:	1.3.11
compile-time dependency versions
	gusb:	0.3.4
	efivar:	37
daemon version:	1.3.11

When I install fwupd via the snap store (i.e. sudo snap install fwupd --classic) on version 1.5.8, the problem for the same device ID is blown away:

├─System Firmware:
│ │   Device ID:          5f177a3bd519ece03c345228af4255a1f5916ae2
│ │   Current version:    65555
│ │   Vendor:             LENOVO (DMI:LENOVO)
│ │   GUIDs:              e1678770-bdc9-47f9-8d90-340e28b6f196
│ │                       230c8b18-8d9b-53ec-838b-6cfc0383493a
│ │                       9d64c9c6-c8e7-569a-b109-2df1038c896b
│ │   Device Flags:       • Internal device
│ │                       • Updatable
│ │                       • System requires external power source
│ │                       • Needs a reboot after installation
│ │                       • Cryptographic hash verification is available
│ │                       • Device is usable for the duration of the update

 But in this case, there is the following warning written:
"WARNING: This package has not been validated, it may not work properly."

So, the summarize the problem: In Version 1.3.11 the TPM PCR0 failure persists. In the most recent version, the problem is gone away, but the version seems not work 100% sure (at least what the warning message implies me).

My questions:
1.) How can I get rid of the TPM PCR0 failure with the a validate version for my ubuntu system and what could be the reason for that problem?
2.) And a more general question, is it important to have a operating system related UEFI installed like in my case? There is still this windows UEFI on my system, but it seems to work properly with my ubuntu operating system. 
3.) I could also upgrade my UEFI in a manual manner with downloading the correct UEFI for ubuntu from https://download.lenovo.com/pccbbs/mobiles/n30ul06w.zip. I've looked into the zip-container and there are two files (N30ET36W.cab (UEFI) and N30HT31W (ECP)). By using the command fwupdmgr install *.cab I could install it manually. Could that approach maybe solve my problems and in which order I have to install the two *.cab files?

Many thanks in advance for the help!

**Comment by hughsie** [comment]
On Thu, 25 Mar 2021 at 11:14, s3venup ***@***.***> wrote:
> It seems it's the UEFI which causes the TPM PCR0 failure.

We fixed this bug a long time ago in fwupd.

> When I install fwupd via the snap store (i.e. sudo snap install fwupd --classic) on version 1.5.8, the problem for the same device ID is blown away:

> But in this case, there is the following warning written:
> "WARNING: This package has not been validated, it may not work properly."

Yep, the snap version has had basically no QA and if it breaks we
don't really support it. The distro versions actually have QA people
testing things, hence the warning.

> By using the command fwupdmgr install *.cab I could install it manually.

"fwupdmgr update" is normally enough to get the updates from the LVFS
from Lenovo, but in this case installing the .cab files manually is
fine too.

Richard.

**Reply by AndreasSchaffhauser** [reply]
Hi Richard,

many thanks for your fast reply! Is there any newer version (1.5.x) what can be installed via apt for ubuntu 20.04.1 focal what resolves that problem and is also validated?

Thanks, Andy.

**Reply by superm1** [reply]
Not for Ubuntu 20.04.  You can install a non-LTS Ubuntu version to get a newer one.  For example Hirsute has 1.5.7.
I would suggest for now you can just ignore the TPM PCR0 warning.

**Comment by hughsie** [comment]
On Thu, 25 Mar 2021 at 11:26, s3venup ***@***.***> wrote:
> many thanks for your fast reply! Is there any newer version (1.5.x) what can be installed via apt what resolves that problem and is also validate?

No idea about Ubuntu I'm afraid, but Fedora 33, 34 and rawhide all
have 1.5.8 builds as of yesterday.

Richard

**Reply by AndreasSchaffhauser** [reply]
One more question in relation to the TPM PCR0 Reconstruction failure. In your wiki there is pointed out, there four possible failures causing this problem:
1.) An error in the firmware TPM event log.
2.) An error in the fwupd reconstruction of the TPM PCR0
3.) A hardware failure
4.) Presence of malware on the system.
How can I rule out that the reason for the error is not point 4 (malware)?

Please find attached the file foo.txt, which contains the /usr/bin/fwupdtpmevlog.

Thanks!
Andy
[foo.txt](https://github.com/fwupd/fwupd/files/6205122/foo.txt)

**Reply by superm1** [reply]
> How can I rule out that the reason for the error is not point 4 (malware)?

With the buggy version of fwupd on the system you won't be able to.  You'll need to have a fwupd version that is known to calculate PCR0 reliably.

**Reply by AndreasSchaffhauser** [reply]
As the problem seems to disappear in version 1.5.8 (installed via snap) for the same Device-ID, can I be sure then? (Please see my first post)

Kind regards,
Andy

**Reply by superm1** [reply]
>As the problem seems to disappear in version 1.5.8 (installed via snap) for the same Device-ID, can I be sure then? (Please see my first post)

Correct.
